evolver/controller/standard/chemostat.py

```diff
@@ -1,66 +1,82 @@
+from collections import defaultdict, deque
 import time
+
 from pydantic import Field
-from collections import defaultdict, deque
+
+from evolver.base import ConfigDescriptor
 from evolver.controller.interface import Controller
-from evolver.hardware.interface import VialConfigBaseModel
+from evolver.hardware.interface import VialConfigBaseModel, HardwareDriver
 
 
 class Chemostat(Controller):
     class Config(VialConfigBaseModel):
-        od_sensor: str = Field(description="name of OD sensor to use")
-        pump: str = Field(description="name of pump device to use")
-        stirrer: str = Field(description="name of stirrer device to use")
+        od_sensor: HardwareDriver | ConfigDescriptor | str = Field(description="name of OD sensor to use")
+        pump: HardwareDriver | ConfigDescriptor | str = Field(description="name of pump device to use")
+        stirrer: HardwareDriver | ConfigDescriptor | str = Field(description="name of stirrer device to use")
         window: int = Field(7, description="number of OD measurements to collect prior to start")
         min_od: float = Field(0, description="OD at which to start chemostat dilutions")
         start_delay: int = Field(0, description="Time (in hours) after which to start dilutions")
         flow_rate: float = Field(0, description="Flow rate for dilutions")
         stir_rate: float = Field(8, description="Stir rate")
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def __init__(self,
+                 *args,
+                 od_sensor: HardwareDriver | ConfigDescriptor | str,
+                 pump: HardwareDriver | ConfigDescriptor | str,
+                 stirrer: HardwareDriver | ConfigDescriptor | str,
+                 **kwargs):
+        self._od_sensor = od_sensor
+        self._pump = pump
+        self._stirrer = stirrer
+
+        # Since ``od_sensor`` and alike are properties that we explicitly initialize above, don't auto assign them
+        # in ``BaseInterface.__init__`` from the ``Config``.
+        super().__init__(*args, auto_config_ignore_fields=("od_sensor", "pump", "stirrer"), **kwargs)
+
         # buffer could come from history as well
-        self.od_buffer = defaultdict(lambda: deque(maxlen=self.config.window))
+        self.od_buffer = defaultdict(lambda: deque(maxlen=self.window))
+
         # start_time is something we might actually want to be a property of
         # evolver, in case of interrupt it has a chance of continuing - but can
         # come from history similarly
         self.start_time = time.time()
 
     @property
-    def od(self):
-        return self.evolver.hardware.get(self.config.od_sensor)
+    def od_sensor(self):
+        return self.evolver.hardware.get(self._od_sensor) if isinstance(self._od_sensor, str) else self._od_sensor
```

Comment:
Wouldn't that just be yet another workaround for prohibiting normal property functionality?

What's the design goal and advantage to prohibiting properties?

Reply:
not to prohibit anything. But the same pattern is repeated 3 times here and perhaps the same pattern would show up in other instances. So it seems it would be quite useful to have something like that

Reply:
>then code user would just self.get_hw(self.od_sensor)

Ah, did you mean for under the hood of the property or instead of the property?

Reply:
If you meant the former, this still kinda defeats the point of having the property here as it's the property's purpose to define this behavior. I think this degree of code duplication is fine.

Reply:
I meant instead of property, then it can be used many places and then don't have to do the special init treatment, it would reduce a lot of code!

Reply:
Oh, ok, that's how I first read then it which means my question still stands since a property seems like a perfect use case here for dynamically retrieving an instance attr.

>What's the design goal and advantage to prohibiting properties?